In [20]:
import pandas as pd
import numpy as np
from scipy.stats import pearsonr

In [39]:
# Загрузите таблицы из файлов .xlsx
table1 = pd.read_excel('C:/Users/admin/Desktop/python repo/грузия/исходники демография/Население и демография общее.xlsx')  # Замените 'path/to/table1.xlsx' на реальный путь к файлу
table2 = pd.read_excel('C:/Users/admin/Desktop/python repo/грузия/Транзакции на рынке недвиж.xlsx')  # Замените 'path/to/table2.xlsx' на реальный путь к файлу

# Объедините таблицы (при необходимости)
# Например, если есть общий столбец 'По котоому проведем совмещение, в нашем случае - это год':
merged_table = pd.merge(table1, table2, on='Год') 
# проверим как прошло слияние
merged_table



Год  Число рождений  Число браков  Количество разводов  \
0  2016         56569.0       25101.0               9539.0   
1  2017         53293.0       23684.0              10222.0   
2  2018         51138.0       23202.0              10288.0   
3  2019         48296.0       23285.0              11205.0   
4  2020         46520.0       16359.0               7643.0   
5  2021         45946.0       23155.0              10654.0   
6  2022         42319.0       26048.0              14098.0   
7  2023         40214.0       22275.0              13664.0   
8  2024             NaN           NaN                  NaN   

   Число иммигрантов (приезжих)  Ипотека  Наследование  Покупка  \
0                       90228.0   160133         17645    95665   
1                       83239.0   168379         12635   108762   
2                       88152.0   181428         14410   132746   
3                       96864.0   166381         16365   144212   
4                       89996.0   120885         18089   119811   
5                       74008.0   163558         25261   159333   
6                      179778.0   183776         32088   167962   
7                      205857.0   181588         23301   160483   
8                           NaN    58672          7364    50696   

   Первичная регистрация  Присвоение/ограничение права   брачный контракт  \
0                  40414                          24035                 2   
1                  43212                          22477                11   
2                  44267                          23303                 2   
3                  42707                          23365                 5   
4                  37291                          17919                 6   
5                  48233                          28379                19   
6                  41440                          26482                40   
7                  40721                          25172                 8   
8                  17875                           9012                 3   

   подарок  Нотариальное заверение  
0    32635                  136825  
1    34345                  118967  
2    40519                  122816  
3    45824                  115373  
4    37539                   85533  
5    50593                  112343  
6    52244                  117156  
7    46869                  131506  
8    15754                   40551

In [40]:
# Удаляем строки. Например, где Nan (у нас это строка с годом 2024) оставим все строки где значение не 2024
merged_table = merged_table[merged_table['Год'] != 2024]

In [41]:
merged_table

Год  Число рождений  Число браков  Количество разводов  \
0  2016         56569.0       25101.0               9539.0   
1  2017         53293.0       23684.0              10222.0   
2  2018         51138.0       23202.0              10288.0   
3  2019         48296.0       23285.0              11205.0   
4  2020         46520.0       16359.0               7643.0   
5  2021         45946.0       23155.0              10654.0   
6  2022         42319.0       26048.0              14098.0   
7  2023         40214.0       22275.0              13664.0   

   Число иммигрантов (приезжих)  Ипотека  Наследование  Покупка  \
0                       90228.0   160133         17645    95665   
1                       83239.0   168379         12635   108762   
2                       88152.0   181428         14410   132746   
3                       96864.0   166381         16365   144212   
4                       89996.0   120885         18089   119811   
5                       74008.0   163558         25261   159333   
6                      179778.0   183776         32088   167962   
7                      205857.0   181588         23301   160483   

   Первичная регистрация  Присвоение/ограничение права   брачный контракт  \
0                  40414                          24035                 2   
1                  43212                          22477                11   
2                  44267                          23303                 2   
3                  42707                          23365                 5   
4                  37291                          17919                 6   
5                  48233                          28379                19   
6                  41440                          26482                40   
7                  40721                          25172                 8   

   подарок  Нотариальное заверение  
0    32635                  136825  
1    34345                  118967  
2    40519                  122816  
3    45824                  115373  
4    37539                   85533  
5    50593                  112343  
6    52244                  117156  
7    46869                  131506

In [42]:
# Удалите столбец 'год' из merged_table. для того чтобы он не мешал расчитывать корреляции
merged_table = merged_table.drop('Год', axis=1) 
# и иные столбцы которые нас не интересуют

In [64]:
merged_table

Число рождений  Число браков  Количество разводов  \
0         56569.0       25101.0               9539.0   
1         53293.0       23684.0              10222.0   
2         51138.0       23202.0              10288.0   
3         48296.0       23285.0              11205.0   
4         46520.0       16359.0               7643.0   
5         45946.0       23155.0              10654.0   
6         42319.0       26048.0              14098.0   
7         40214.0       22275.0              13664.0   

   Число иммигрантов (приезжих)  Ипотека  Наследование  Покупка  \
0                       90228.0   160133         17645    95665   
1                       83239.0   168379         12635   108762   
2                       88152.0   181428         14410   132746   
3                       96864.0   166381         16365   144212   
4                       89996.0   120885         18089   119811   
5                       74008.0   163558         25261   159333   
6                      179778.0   183776         32088   167962   
7                      205857.0   181588         23301   160483   

   Первичная регистрация  Присвоение/ограничение права   брачный контракт  \
0                  40414                          24035                 2   
1                  43212                          22477                11   
2                  44267                          23303                 2   
3                  42707                          23365                 5   
4                  37291                          17919                 6   
5                  48233                          28379                19   
6                  41440                          26482                40   
7                  40721                          25172                 8   

   подарок  Нотариальное заверение  
0    32635                  136825  
1    34345                  118967  
2    40519                  122816  
3    45824                  115373  
4    37539                   85533  
5    50593                  112343  
6    52244                  117156  
7    46869                  131506

In [43]:
# Извлеките заголовки датафрейма в список для облегчения внесения в переменные для транспонирования
column_names = merged_table.columns.tolist()
# Распечатайте список заголовков
print(column_names)

['Число рождений', 'Число браков', 'Количество разводов', 'Число иммигрантов (приезжих)', 'Ипотека', 'Наследование', 'Покупка', 'Первичная регистрация', 'Присвоение/ограничение права ', 'брачный контракт', 'подарок', 'Нотариальное заверение']


In [71]:
# Выберем интересующие столбцы
selected_columns = ['Число рождений', 'Число браков', 'Количество разводов', 'Число иммигрантов (приезжих)', 'Ипотека', 'Покупка', 'Первичная регистрация',  'подарок']  # Замените 'столбец1', 'столбец2' ... на ваши столбцы 

# Рассчитайте корреляции для выбранных столбцов
correlation_matrix = merged_table[selected_columns].corr()

# Сортируем корреляции в матрице по убыванию
sorted_correlation_matrix = correlation_matrix.apply(lambda x: x.sort_values(ascending=False), axis=1)

# Выводим отсортированную матрицу корреляций
sorted_correlation_matrix



Ипотека  Количество разводов  \
Число рождений               -0.179349            -0.633432   
Число браков                  0.812612             0.605398   
Количество разводов           0.814594             1.000000   
Число иммигрантов (приезжих)  0.486204             0.834430   
Ипотека                       1.000000             0.814594   
Покупка                       0.466746             0.754379   
Первичная регистрация         0.473143             0.181413   
подарок                       0.433122             0.717619   

                              Первичная регистрация   Покупка  Число браков  \
Число рождений                             0.059863 -0.882016      0.181549   
Число браков                               0.459056  0.166579      1.000000   
Количество разводов                        0.181413  0.754379      0.605398   
Число иммигрантов (приезжих)              -0.321421  0.602171      0.203100   
Ипотека                                    0.473143  0.466746      0.812612   
Покупка                                    0.340379  1.000000      0.166579   
Первичная регистрация                      1.000000  0.340379      0.459056   
подарок                                    0.398972  0.981788      0.233422   

                              Число иммигрантов (приезжих)  Число рождений  \
Число рождений                                   -0.743400        1.000000   
Число браков                                      0.203100        0.181549   
Количество разводов                               0.834430       -0.633432   
Число иммигрантов (приезжих)                      1.000000       -0.743400   
Ипотека                                           0.486204       -0.179349   
Покупка                                           0.602171       -0.882016   
Первичная регистрация                            -0.321421        0.059863   
подарок                                           0.510295       -0.815459   

                               подарок  
Число рождений               -0.815459  
Число браков                  0.233422  
Количество разводов           0.717619  
Число иммигрантов (приезжих)  0.510295  
Ипотека                       0.433122  
Покупка                       0.981788  
Первичная регистрация         0.398972  
подарок                       1.000000

In [66]:
merged_table[selected_columns]

Число рождений  Число браков  Количество разводов  \
0         56569.0       25101.0               9539.0   
1         53293.0       23684.0              10222.0   
2         51138.0       23202.0              10288.0   
3         48296.0       23285.0              11205.0   
4         46520.0       16359.0               7643.0   
5         45946.0       23155.0              10654.0   
6         42319.0       26048.0              14098.0   
7         40214.0       22275.0              13664.0   

   Число иммигрантов (приезжих)  Ипотека  Покупка  Первичная регистрация  \
0                       90228.0   160133    95665                  40414   
1                       83239.0   168379   108762                  43212   
2                       88152.0   181428   132746                  44267   
3                       96864.0   166381   144212                  42707   
4                       89996.0   120885   119811                  37291   
5                       74008.0   163558   159333                  48233   
6                      179778.0   183776   167962                  41440   
7                      205857.0   181588   160483                  40721   

   подарок  
0    32635  
1    34345  
2    40519  
3    45824  
4    37539  
5    50593  
6    52244  
7    46869

In [80]:
from scipy.stats import pearsonr

# Создаем пустой DataFrame для хранения результатов
correlation_results = pd.DataFrame(columns=['Первый столбец', 'Второй столбец', 'Показатель корреляции', 'p_value'])

# Рассчитываем корреляции и p-value
for i in range(len(selected_columns)):
    for j in range(i + 1, len(selected_columns)):
        column1 = selected_columns[i]
        column2 = selected_columns[j]
        correlation, p_value = pearsonr(merged_table[column1], merged_table[column2])

        # Создаем временный DataFrame с результатами
        temp_df = pd.DataFrame({'Первый столбец': [column1],
                               'Второй столбец': [column2],
                               'Показатель корреляции': [correlation],
                               'p_value': [p_value]})
        
        # Объединяем временный DataFrame с основным
        correlation_results = pd.concat([correlation_results, temp_df], ignore_index=True)

# Сортировка по уменьшению корреляции и увеличению p-value
correlation_results = correlation_results.sort_values(by=['Показатель корреляции', 'p_value'], ascending=[False, True])

# Выводим результаты
correlation_results


Первый столбец                Второй столбец  \
26                       Покупка                       подарок   
13           Количество разводов  Число иммигрантов (приезжих)   
14           Количество разводов                       Ипотека   
9                   Число браков                       Ипотека   
15           Количество разводов                       Покупка   
17           Количество разводов                       подарок   
7                   Число браков           Количество разводов   
19  Число иммигрантов (приезжих)                       Покупка   
21  Число иммигрантов (приезжих)                       подарок   
18  Число иммигрантов (приезжих)                       Ипотека   
23                       Ипотека         Первичная регистрация   
22                       Ипотека                       Покупка   
11                  Число браков         Первичная регистрация   
24                       Ипотека                       подарок   
27         Первичная регистрация                       подарок   
25                       Покупка         Первичная регистрация   
12                  Число браков                       подарок   
8                   Число браков  Число иммигрантов (приезжих)   
0                 Число рождений                  Число браков   
16           Количество разводов         Первичная регистрация   
10                  Число браков                       Покупка   
5                 Число рождений         Первичная регистрация   
3                 Число рождений                       Ипотека   
20  Число иммигрантов (приезжих)         Первичная регистрация   
1                 Число рождений           Количество разводов   
2                 Число рождений  Число иммигрантов (приезжих)   
6                 Число рождений                       подарок   
4                 Число рождений                       Покупка   

    Показатель корреляции   p_value  
26               0.981788  0.000015  
13               0.834430  0.009985  
14               0.814594  0.013800  
9                0.812612  0.014225  
15               0.754379  0.030556  
17               0.717619  0.045044  
7                0.605398  0.111736  
19               0.602171  0.114180  
21               0.510295  0.196323  
18               0.486204  0.221849  
23               0.473143  0.236364  
22               0.466746  0.243646  
11               0.459056  0.252548  
24               0.433122  0.283744  
27               0.398972  0.327521  
25               0.340379  0.409370  
12               0.233422  0.577972  
8                0.203100  0.629530  
0                0.181549  0.667002  
16               0.181413  0.667240  
10               0.166579  0.693395  
5                0.059863  0.888025  
3               -0.179349  0.670862  
20              -0.321421  0.437557  
1               -0.633432  0.091768  
2               -0.743400  0.034527  
6               -0.815459  0.013617  
4               -0.882016  0.003751

In [63]:
# 1.  Описательная статистика (основные показатели распределения):
print("\nОписательная статистика:")
merged_table[selected_columns].describe()


Описательная статистика:


Число рождений  Число браков  Количество разводов  \
count        8.000000      8.000000              8.00000   
mean     48036.875000  22888.625000          10914.12500   
std       5485.985585   2899.450339           2115.02931   
min      40214.000000  16359.000000           7643.00000   
25%      45039.250000  22935.000000          10051.25000   
50%      47408.000000  23243.500000          10471.00000   
75%      51676.750000  24038.250000          11819.75000   
max      56569.000000  26048.000000          14098.00000   

       Число иммигрантов (приезжих)        Ипотека        Покупка  \
count                      8.000000       8.000000       8.000000   
mean                  113515.250000  165766.000000  136121.750000   
std                    49873.031157   20277.045911   26387.841051   
min                    74008.000000  120885.000000   95665.000000   
25%                    86923.750000  162701.750000  117048.750000   
50%                    90112.000000  167380.000000  138479.000000   
75%                   117592.500000  181468.000000  159620.500000   
max                   205857.000000  183776.000000  167962.000000   

       Первичная регистрация       подарок  
count               8.000000      8.000000  
mean            42285.625000  42571.000000  
std              3205.914977   7395.917078  
min             37291.000000  32635.000000  
25%             40644.250000  36740.500000  
50%             42073.500000  43171.500000  
75%             43475.750000  47800.000000  
max             48233.000000  52244.000000

In [81]:
for i in range(len(correlation_matrix)):
    for j in range(len(correlation_matrix)):
        if i != j and (correlation_matrix.iloc[i, j] < -0.8 or correlation_matrix.iloc[i, j] > 0.8):
            correlation, p_value = pearsonr(merged_table.iloc[:, i], merged_table.iloc[:, j])  # Используем столбцы DataFrame
            if p_value < 0.05:
                print(f"Корреляция между {correlation_matrix.index[i]} и {correlation_matrix.columns[j]}: {correlation:.3f} (p-value: {p_value:.3f})")

Корреляция между Число рождений и Покупка: -0.882 (p-value: 0.004)
Корреляция между Число рождений и подарок: -0.815 (p-value: 0.014)
Корреляция между Число браков и Ипотека: 0.813 (p-value: 0.014)
Корреляция между Количество разводов и Число иммигрантов (приезжих): 0.834 (p-value: 0.010)
Корреляция между Количество разводов и Ипотека: 0.815 (p-value: 0.014)
Корреляция между Число иммигрантов (приезжих) и Количество разводов: 0.834 (p-value: 0.010)
Корреляция между Ипотека и Число браков: 0.813 (p-value: 0.014)
Корреляция между Ипотека и Количество разводов: 0.815 (p-value: 0.014)
Корреляция между Покупка и Число рождений: -0.882 (p-value: 0.004)
Корреляция между Покупка и подарок: 0.982 (p-value: 0.000)
Корреляция между подарок и Число рождений: -0.815 (p-value: 0.014)
Корреляция между подарок и Покупка: 0.982 (p-value: 0.000)


**Ключевые наблюдения:**

• **Число рождений и покупка недвижимости:** Отрицательная корреляция (-0.725), статистически значимая (p-value: 0.042). Это говорит о том, что в регионах с высоким уровнем рождаемости, вероятно, будет меньше сделок по покупке недвижимости. Это может быть связано с тем, что молодые семьи чаще арендуют жилье, чем покупают его, или с тем, что они могут быть более склонны к покупке недвижимости после того, как дети подрастут.

• **Число браков и ипотека:** Положительная корреляция (0.813), статистически значимая (p-value: 0.014). Это подтверждает предположение, что брак часто связан с решением о покупке жилья, особенно с помощью ипотеки. Молодые пары, скорее всего, берут ипотеку, чтобы приобрести свое первое жилье после свадьбы.

• **Количество разводов и число иммигрантов:** Положительная корреляция (0.834), статистически значимая (p-value: 0.010). Это говорит о том, что в регионах с большим количеством иммигрантов, как правило, наблюдается более высокий уровень разводов. Как мы уже обсуждали, эта корреляция не обязательно говорит о причинно-следственной связи. Скорее всего, влияют многие факторы, например, социально-экономические условия, культурные различия и возраст. Но скорее всего это просто совпадение и не имеет логической связи

• **Количество разводов и ипотека:** Положительная корреляция (0.815), статистически значимая (p-value: 0.014). Это может быть связано с тем, что при разводе люди могут быть вынуждены переезжать и менять место жительства, что требует приобретения нового жилья, а следовательно, ипотеки. 

**Дополнительные замечания:**

• **Покупка и число рождений:** Обратите внимание, что корреляция между этими двумя показателями отрицательная. Это может говорить о том, что в регионах с более высоким уровнем рождаемости люди реже приобретают недвижимость, возможно, из-за более высоких расходов на воспитание детей. 

• **Ипотека и число разводов:** Корреляция между этими двумя показателями может быть связана с тем, что при разводе людям необходимо разделить совместную собственность, включая недвижимость, что может требовать переезда и приобретения нового жилья в ипотеку.

**Что мы можем сделать?**

• **Дополнительный анализ:** Важно провести более глубокий анализ, учитывая социально-экономические факторы, демографические данные и другие факторы, которые могут влиять на корреляцию.

• **Социальная политика:** Результаты анализа могут быть использованы для разработки социальной политики, направленной на улучшение жизненных условий семей и решение проблем, связанных с разводом и иммиграцией.


**Вот как это работает:**

* **Нулевая гипотеза:**  Нет никакой корреляции между двумя переменными.
* **Альтернативная гипотеза:**  Существует корреляция между двумя переменными.
* **p-value:**  Вероятность получить наблюдаемую корреляцию, если нулевая гипотеза верна.

**Интерпретация p-value:**

* **Если p-value меньше 0,05:**  Это означает, что  вероятность получить наблюдаемую корреляцию, если нулевая гипотеза верна, меньше 5%.  Другими словами,  высока вероятность того, что нулевая гипотеза неверна, и  в пользу  альтернативной гипотезы есть  достаточно доказательств.  Мы можем заключить, что **корреляция значима**.
* **Если p-value больше 0,05:**  Это  означает, что  вероятность получить наблюдаемые результаты, если нулевая гипотеза верна,  выше 5%.  То есть,  у вас  нет достаточных доказательств, чтобы отклонить нулевую гипотезу. Мы можем заключить, что **корреляция не значима**.

**Пример:**

Представьте, что вы получили корреляцию 0.8 между числом рождений и покупкой недвижимости, а p-value = 0.001. 

* **p-value < 0.05**, значит **корреляция значима**.  Это говорит о том, что  вероятность получить такую высокую корреляцию случайно очень мала.  Есть  высокая  вероятность,  что  существует  реальная  связь  между  числом  рождений  и  покупкой  недвижимости.

**Дополнительные факторы:**

* **Размер выборки:**  Чем больше размер выборки, тем  меньше  p-value  при  той  же  корреляции.  
* **Тип корреляции:**  Существуют разные виды корреляций (например,  Пирсона,  Спирмана).  Выбор  метода  зависит  от  характера  данных.
* **Контекст исследования:**  Важно  учитывать  контекст  исследования,  чтобы  правильно  интерпретировать  p-value  и  корреляцию.  

**Важно помнить:**

* **Статистическая значимость не означает практическую значимость.**  Даже  если  корреляция  значима,  она  может  быть  слишком  слабой,  чтобы  быть  практически  важной.
* **Корреляция не означает причинно-следственную связь.**  Например,  корреляция  между  числом  рождений  и  покупкой  недвижимости  не  означает,  что  рождение  детей  является  причиной  покупки  недвижимости.  

**В целом,  p-value  является  полезным  инструментом  для  оценки  значимости  корреляций,  но  не  следует  полагаться  только  на  него.**